lib/services/android-plugin-build-service.ts

```diff
@@ -172,23 +173,79 @@ export class AndroidPluginBuildService implements IAndroidPluginBuildService {
 		this.validateOptions(options);
 		const manifestFilePath = this.getManifest(options.platformsAndroidDirPath);
 		const androidSourceDirectories = this.getAndroidSourceDirectories(options.platformsAndroidDirPath);
-		const shouldBuildAar = !!manifestFilePath || androidSourceDirectories.length > 0;
+		const shortPluginName = getShortPluginName(options.pluginName);
+		const pluginTempDir = path.join(options.tempPluginDirPath, shortPluginName);
+		const pluginSourceFileHashesInfo = await this.getSourceFilesHashes(options.platformsAndroidDirPath, shortPluginName);
+
+		const shouldBuildAar = await this.shouldBuildAar({
+			manifestFilePath,
+			androidSourceDirectories,
+			pluginTempDir,
+			pluginSourceDir: options.platformsAndroidDirPath,
+			shortPluginName,
+			fileHashesInfo: pluginSourceFileHashesInfo
+		});
 
 		if (shouldBuildAar) {
-			const shortPluginName = getShortPluginName(options.pluginName);
-			const pluginTempDir = path.join(options.tempPluginDirPath, shortPluginName);
-			const pluginTempMainSrcDir = path.join(pluginTempDir, "src", "main");
+			// In case plugin was already built in the current process, we need to clean the old sources as they may break the new build.
+			this.$fs.deleteDirectory(pluginTempDir);
+			this.$fs.ensureDirectoryExists(pluginTempDir);
 
+			const pluginTempMainSrcDir = path.join(pluginTempDir, "src", "main");
 			await this.updateManifest(manifestFilePath, pluginTempMainSrcDir, shortPluginName);
 			this.copySourceSetDirectories(androidSourceDirectories, pluginTempMainSrcDir);
 			await this.setupGradle(pluginTempDir, options.platformsAndroidDirPath, options.projectDir);
 			await this.buildPlugin({ pluginDir: pluginTempDir, pluginName: options.pluginName });
 			this.copyAar(shortPluginName, pluginTempDir, options.aarOutputDir);
+			this.writePluginHashInfo(pluginSourceFileHashesInfo, pluginTempDir);
 		}
 
 		return shouldBuildAar;
 	}
 
+	private getSourceFilesHashes(pluginTempPlatformsAndroidDir: string, shortPluginName: string): Promise<IStringDictionary> {
+		const pathToAar = path.join(pluginTempPlatformsAndroidDir, `${shortPluginName}.aar`);
+		const pluginNativeDataFiles = this.$fs.enumerateFilesInDirectorySync(pluginTempPlatformsAndroidDir, (file: string, stat: IFsStats) => {
+			return file !== pathToAar;
+		});
+
+		return this.$filesHashService.generateHashes(pluginNativeDataFiles);
+	}
+
+	private async writePluginHashInfo(fileHashesInfo: IStringDictionary, pluginTempDir: string): Promise<void> {
+		const buildDataFile = this.getPathToPluginBuildDataFile(pluginTempDir);
+		this.$fs.writeJson(buildDataFile, fileHashesInfo);
+	}
+
+	private async shouldBuildAar(opts: {
+		manifestFilePath: string,
+		androidSourceDirectories: string[],
+		pluginTempDir: string,
+		pluginSourceDir: string,
+		shortPluginName: string,
+		fileHashesInfo: IStringDictionary
+	}): Promise<boolean> {
+
+		let shouldBuildAar = !!opts.manifestFilePath || opts.androidSourceDirectories.length > 0;
```

Comment:
In this way `shouldBuildAar` will become integer instead of boolean. I suppose that even the tslint check will fail as the method should return boolean. I believe that we should not change the variable types based on some conditions. If we want this one shorter, we could use `let shouldBuildAar = !!opts.manifestFilePath || !!opts.androidSourceDirectories.length;` if you believe that its more readable.